### ISSUE TITLE
op_ref false positive

### ISSUE THREAD

**Issue by euclio** [rust-lang-nursery/rust-clippy#2042]
With the following code: https://play.rust-lang.org/?gist=1aa5789d7082019e91bac7b093304ad5&version=stable

Clippy suggests removing the reference, but without the reference the Rust compiler complains that type annotations are needed.

I would expect the lint to not fire in this case.

**Comment by oli-obk** [comment]
That is so weird. I think this is rather a rustc bug... I'll investigate

**Comment by arielb1** [comment]
That's a trait-system issue. Operator overloading does succeed in this case. Without the autoref, you get these trait bounds:
```Rust
u16: Add<$0>
u8: Into<$0>
```

Because the compiler only considers each trait bound alone, it can't see that the only solution is `$0 = u16`.

If you add a reference, instead yoou get
```Rust
u16: Add<&'a $0>
u8: Into<$0>
```

And on the first bound, only the `impl Add<&u16> for u16` impl matches, which allows a solution.

**Comment by samueltardieu** [comment]
For reference, `-Znext-solver` gives a more precise diagnostic about the eligible trait implementations but still requires an explicit annotation:

```rust
error[E0283]: type annotations needed
 --> /tmp/t.rs:5:25
  |
5 |     let x = word + byte.into();
  |                  -      ^^^^
  |                  |
  |                  type must be known at this point
  |
  = note: multiple `impl`s satisfying `u16: std::ops::Add<_>` found in the `core` crate:
          - impl std::ops::Add for u16;
          - impl std::ops::Add<&u16> for u16;
help: try using a fully qualified path to specify the expected types
  |
5 |     let x = word + <u8 as std::convert::Into<T>>::into(byte);
  |                    ++++++++++++++++++++++++++++++++++++    ~
```
